lib/topological_inventory/orchestrator/event_manager.rb

```diff
@@ -0,0 +1,128 @@
+module TopologicalInventory
+  module Orchestrator
+    # Event Manager subscribes to kafka topic "platform.sources.event-stream"
+    # and listens to messages from Sources API: create/update/destroy
+    # on models used for computing of digest (see Source#digest_values)
+    #
+    # When event is received then sync API -> OpenShift is invoked.
+    # If no event is caught for >= 1 hour, sync is invoked too.
+    class EventManager
+      include Logging
+
+      SYNC_EVENT_INTERVAL = 1.hour
+      SKIP_SUBSEQUENT_EVENTS_DURATION = 5.seconds
+
+      def self.run!(worker)
+        manager = new(worker)
+        manager.run!
+      end
+
+      def initialize(worker)
+        self.worker = worker
+        self.event_semaphore = Mutex.new
+        self.sync_semaphore  = Mutex.new
```

Comment:
I'd really like to avoid coordinating a multi-mutex dance between threads if at all possible

If we drop the "only do a full sync if there wasn't an event in the last hour" then we don't have to synchronize on last_event_at.

And if you had only a single thread doing the sync openshift with sources then you don't need the second one.  This could be done a few ways but one way could be to have the sync thread listen to an in-memory queue that the kafka thread or the scheduled_sync thread push to.  This is how most of our MIQ event catchers work.

Reply:
well, I'm, not sure if I understand. 
1) `last_event_at` is there primarily for skipping multievents generated by Sources API. 
2) if we drop the no-event sync, then we are actually saying: "I believe nothing wrong in another application can happen. Never"...emm, why?
3) both mutexes have different usage, `event_semaphore` sets last sync and handles multievents, `sync_semaphore` handles events from different sources. 
4) code isn't wrong only because it uses threads :)

Reply:
> 1. last_event_at is there primarily for skipping multievents generated by Sources API.

That's fine, I'm saying accessing last_event_at from both threads is what is making that mutex necessary

> 2. if we drop the no-event sync, then we are actually saying: "I believe nothing wrong in another application can happen. Never"...emm, why?

Huh?  That's not what I'm saying at all.

> 3. both mutexes have different usage, event_semaphore sets last sync and handles multievents, sync_semaphore handles events from different sources.

Yeah I got that, I'm saying that they might not be necessary if we structure the code differently

> 4. code isn't wrong only because it uses threads :)

That's also not what I'm saying.  Sharing variables between threads is however more complicated and more likely that in 2 years someone who doesn't understand threading as well will screw it up.

In fact I think _adding_ a thread but removing any shared variables between threads would be better